File: interviewBit-golang/src/arrays/problems/spiralMatrix.go
package arrayProblems

/*
	Problem : https://www.interviewbit.com/problems/spiral-matrix/
*/

// T(n) : O(n), S(n) : O(1)
func SpiralMatrix(A []int, B int, C int) [][]int {

	res := make([][]int, 0)
	for i := 0; i < B; i++ {
		arr := make([]int, C)
		res = append(res, arr)
	}

	i, n := 0, len(A)
	rs, re, cs, ce := 0, B-1, 0, C-1
	for i < n {
		for j := cs; j <= ce && i < n; j++ {
			res[rs][j] = A[i]
			i++
		}
		rs++

		for j := rs; j <= re && i < n; j++ {
			res[j][ce] = A[i]
			i++
		}
		ce--

		for j := ce; j >= cs && i < n; j-- {
			res[re][j] = A[i]
			i++
		}
		re--

		for j := re; j >= rs && i < n; j-- {
			res[j][cs] = A[i]
			i++
		}
		cs++
	}

	return res
}

File: interviewBit-cpp/src/arrays/spiral-matrix.cpp
#include <iostream>
#include <vector>

using namespace std;

// https://www.interviewbit.com/problems/spiral-matrix/

// Approach 1
// T(n) : O(n) ; S(n) : O(n)
vector<vector<int> > spiralMatrix(vector<int> &A, int B, int C) {

    vector<vector<int> > spiral(B, vector<int>(C, 0));
    int rs = 0, re = B-1, cs = 0, ce = C-1, idx = 0;
    while (rs <= re && cs <= ce) {

        for (int i = cs; i <= ce; i++) {
            spiral[rs][i] = A[idx++];
        }
        rs++;
        if (cs > ce) break;

        for (int i = rs; i <= re; i++) {
            spiral[i][ce] = A[idx++];
        }
        ce--;
        if (rs > re) break;

        for (int i = ce; i >= cs; i--) {
            spiral[re][i] = A[idx++];
        }
        re--;
        if (cs > ce) break;

        for (int i = re; i >= rs; i--) {
            spiral[i][cs] = A[idx++];
        }
        cs++;
    }

    return spiral;
}

// Driver Code for testing
int main() {
    vector<int> arr = {1,2,3,4};
    vector<vector<int> > mat = spiralMatrix(arr, 2,  2);

    for (auto &vec : mat) {
        for (auto &v : vec) {
            cout << v << " ";
        }
        cout << "\n";
    }

    vector<vector<int> > mat1 = spiralMatrix(arr, 1, 4);
    for (auto &vec : mat1) {
        for (auto &v : vec) {
            cout << v << " ";
        }
        cout << "\n";
    }

    return 0;
}
